framework/validators/IpValidator.php

```diff
@@ -0,0 +1,555 @@
+            }
+
+            if (!$this->ipv6) {
+                return [$this->ipv6NotAllowed, []];
+            }
+            if (!$this->validate6($ip)) {
+                return [$this->wrongIp, []];
+            }
+
+            if ($this->expandV6) {
+                $ip = $this->expand6($ip);
+            }
+        } else {
+            if ($cidr !== null) {
+                if ($cidr > static::IPV4_ADDRESS_LENGTH || $cidr < 0) {
+                    return [$this->wrongCidr, []];
+                }
+            } else {
+                $isCidrDefault = true;
+                $cidr = static::IPV4_ADDRESS_LENGTH;
+            }
+
+            if (!$this->ipv4) {
+                return [$this->ipv4NotAllowed, []];
+            }
+            if (!$this->validate4($ip)) {
+                return [$this->wrongIp, []];
+            }
+        }
+
+        if (!$this->isAllowed($ip, $cidr)) {
+            return [$this->notInRange, []];
+        }
+
+        $result = $negation . $ip;
+
+        if ($this->subnet !== false && (!$isCidrDefault || $isCidrDefault && $this->normalize)) {
+            $result .= "/$cidr";
+        }
+
+        return $result;
+    }
+
+    /**
+     * Expands an IPv6 address to it's full notation. For example `2001:db8::1` will be
+     * expanded to `2001:0db8:0000:0000:0000:0000:0000:0001`
+     *
+     * @param string $ip the original IPv6
+     * @return string the expanded IPv6
+     */
+    public static function expand6($ip)
+    {
+        $hex = unpack("H*hex", inet_pton($ip));
+        return substr(preg_replace("/([a-f0-9]{4})/i", "$1:", $hex['hex']), 0, -1);
+    }
+
+    /**
+     * The method checks, if the IP address is allowed according to
+     * the [[allow]] and [[deny]] lists, and the [[order]] property.
+     *
+     * @param string $ip
+     * @param integer $cidr
+     * @return boolean
+     * @see order
+     */
+    public function isAllowed($ip, $cidr)
+    {
+        $denied = false;
+        $allowed = true;
+        if (!empty($this->deny) && $this->inRange($ip, $cidr, $this->deny)) {
+            $denied = true;
+        }
+        if (!empty($this->allow) && !$this->inRange($ip, $cidr, $this->allow)) {
+            $allowed = false;
+        }
+
+        return $this->order === self::ORDER_DENY_ALLOW ? (!$denied && $allowed) : (!$denied || $allowed);
+    }
+
+    /**
+     * Validates IPv4 address
+     *
+     * @param string $value
+     * @return boolean
+     */
+    public function validate4($value)
+    {
+        return preg_match($this->ipv4Pattern, $value) !== 0;
+    }
+
+    /**
+     * Validates IPv6 address
+     *
+     * @param string $value
+     * @return boolean
+     */
+    public function validate6($value)
+    {
+        return preg_match($this->ipv6Pattern, $value) !== 0;
+    }
+
+    /**
+     * Gets the IP version
+     *
+     * @param string $ip
+     * @return integer
+     */
+    public static function getIpVersion($ip)
```

Comment:
It's public and static. It may be even in a helper, but we don't need a separated helper to work with IP addresses.